### ISSUE TITLE
Adding encode_corpus and encode_queries to SentenceTransformers

### ISSUE THREAD

**Issue by KennethEnevoldsen** [UKPLab/sentence-transformers#2776]
From MTEB we have for a while used `encode_corpus` and `encode_queries` functions to allow for conditional encodes based on the type of information encoded.

**Using prompt_name is not enough**:
For a model like e5 instruct we see typically want the following pattern:
```
if encode_queries:
  
  model.encode(sentences, prompt_name=prompt_name)
  
# assuming the model implements the following function for encoding the corpus:
def get_detailed_instruct(task_description: str, sentence: str) -> str:
  return f'Instruct: {task_description}\nQuery: {sentence}'
```

However, if we want to use it for encoding the corpus we get:

```
if encode_corpus:
  
  model.encode(sentences, prompt_name)
  
# Assuming the model implements the following function for encoding the corpus:
def get_detailed_instruct(task_description: str, sentence: str) -> str:
  return "sentence"
```

These are implementation details that I would love to encode in the model object, but as it is set up now it seems like you are required to modify or wrap the model object to do so.

**Alternative solution**:

An alternative solution is to add an optional "encode_type" argument to the encode function.

@tomaarsen let me know what you think about this. I am more than willing to create a PR if needed.

**Comment by dbuades** [comment]
I agree that this would be very useful, since we are facing the same issue on our side. I'm happy to contribute too !

@tomaarsen do you have any thoughts on this?

Thank you!

**Comment by KennethEnevoldsen** [comment]
Sounds great. Looking forward to seeing the approach

**Comment by KennethEnevoldsen** [comment]
@dbuades never heard back from you re. the workaround - any updates?

**Comment by dbuades** [comment]
> @dbuades never heard back from you re. the workaround - any updates?

Hey @KennethEnevoldsen thanks for the heads-up. The changes work on our side, I just need to adapt the tests before opening a PR. I wanted to do it last week but other priorities got in the way.

I'll open it this week for sure.

**Comment by KennethEnevoldsen** [comment]
Wonderful. Looking forward to it!

**Comment by dbuades** [comment]
> Wonderful. Looking forward to it!

PR is up [here](https://github.com/embeddings-benchmark/mteb/pull/1221). Looking forward to your thoughts and comments!

**Comment by tomaarsen** [comment]
@KennethEnevoldsen @dbuades 

I've decided I will pick this up in the v5.0 release. They'll be called `encode_query` and `encode_document`, and they'll contain two main changes compared to the regular `model.encode`:
1. If it exists, the "query" or "document" predefined prompt will be used.
2. If the model is a "Router", i.e. a model with multiple routes depending on the `task_type`, then the "query" or "document" routes will be used. This is meant for asymmetric models, a type of model that'll be better supported for both training & inference in v5.0 onwards.

- Tom Aarsen

**Comment by KennethEnevoldsen** [comment]
Great to hear @tomaarsen! (@Samoed just letting you know as well)

**Comment by dbuades** [comment]
> @KennethEnevoldsen @dbuades 
> 
> I've decided I will pick this up in the v5.0 release. They'll be called `encode_query` and `encode_document`, and they'll contain two main changes compared to the regular `model.encode`:
> 1. If it exists, the "query" or "document" predefined prompt will be used.
> 2. If the model is a "Router", i.e. a model with multiple routes depending on the `task_type`, then the "query" or "document" routes will be used. This is meant for asymmetric models, a type of model that'll be better supported for both training & inference in v5.0 onwards.
> 
> - Tom Aarsen

Thanks for letting us know @tomaarsen ! Eager to try it out.

**Comment by tomaarsen** [comment]
Now that the v5.0 is out, I think we can close this. 
Documentation:
* [encode](https://sbert.net/docs/package_reference/sentence_transformer/SentenceTransformer.html#sentence_transformers.SentenceTransformer.encode)
* [encode_query](https://sbert.net/docs/package_reference/sentence_transformer/SentenceTransformer.html#sentence_transformers.SentenceTransformer.encode_query)
* [encode_document](https://sbert.net/docs/package_reference/sentence_transformer/SentenceTransformer.html#sentence_transformers.SentenceTransformer.encode_document)

Thank you again for the suggestion!

- Tom Aarsen
